engine/consensus/dkg/processor/engine.go

```diff
@@ -136,36 +125,70 @@ func (e *Engine) Process(originID flow.Identifier, event interface{}) error {
 func (e *Engine) process(originID flow.Identifier, event interface{}) error {
 	switch v := event.(type) {
 	case msg.DKGMessage:
-		return e.onMessage(originID, v)
+		return e.onPrivateMessage(originID, v)
 	default:
 		return fmt.Errorf("invalid event type (%T)", event)
 	}
 }
 
-func (e *Engine) onMessage(originID flow.Identifier, msg msg.DKGMessage) error {
-	if currentDKG := e.GetDKGInstanceID(); currentDKG != msg.DKGInstanceID {
-		return fmt.Errorf("wrong DKG instance ID. Got %v, want %v", msg.DKGInstanceID, currentDKG)
-	}
-	if msg.Orig >= len(e.committee) || msg.Orig < 0 {
-		return fmt.Errorf("origin id out of range: %d", msg.Orig)
+func (e *Engine) onPrivateMessage(originID flow.Identifier, msg msg.DKGMessage) error {
+	err := e.checkMessageInstanceAndOrigin(msg)
+	if err != nil {
+		return err
 	}
+	// check that the message's origin matches the sender's flow identifier
 	nodeID := e.committee[msg.Orig]
 	if !bytes.Equal(nodeID[:], originID[:]) {
 		return fmt.Errorf("OriginID (%v) does not match committee member %d (%v)", originID, msg.Orig, nodeID)
 	}
-	e.log.Debug().Msgf("forwarding Message to controller")
+	e.log.Debug().Msgf("forwarding private message to controller")
 	e.msgCh <- msg
 	return nil
 }
 
+// fetchBroadcastMessages calls the DKG smart contract to get missing DKG
+// messages for the current epoch, and forwards them to the msgCh.
+// It should be called with the ID of block whose seal is finalized.
+func (e *Engine) fetchBroadcastMessages(referenceBlock flow.Identifier) error {
+	msgs, err := e.dkgContractClient.ReadBroadcast(e.messageOffset, referenceBlock)
+	if err != nil {
+		return fmt.Errorf("could not read broadcast messages: %w", err)
+	}
+	for _, msg := range msgs {
+		err := e.checkMessageInstanceAndOrigin(msg)
+		if err != nil {
+			e.log.Error().Err(err).Msg("bad broadcast message")
+			continue
+		}
+		e.log.Debug().Msgf("forwarding broadcast message to controller")
+		e.msgCh <- msg
+	}
+	e.messageOffset += uint(len(msgs))
```

Comment:
Related to the comment above: we need to synchronize the `messageOffset`  here as well. If we are adding the number of messages received, probably need to lock around this whole function.

Reply:
Yes good point. I left that out for now as I think it will fall into place when we implement the component that calls `fetchBroadcastMessages`. I am not entirely sure how it will control the engine's `dkgInstanceID` and `messageOffset` but I am making a not of your comment.